README
Postfilter 0.9.1 - README


1. Introduction

	postfilter is an antispam filter for InterNetNews, a widely used NNTP server. Unlike cleanfeed - which checks *remote* articles, those 
	messages sent through other sites - postfilter scans *local* posts and tries to reject spam and various abuses. Even if it's designed for 
	large sites, notably USENET servers, postfilter *can* be safety used also by small or private hosts that need a less strong protection 
	against abuses. 

	Beware! Postfilter can't be used as a cleanfeed replacement: cleanfeed checks *remote* articles, postfilter *local* ones.

2. Requirements

	2.1 INN >= 2.4.0 compiled with perl support (--with-perl). It's recommended to use INN >= 2.4.4. 

	Beware! Postfilter could be *not* fully compatible with IPv6 at the moment. 

	2.2 The following perl modules:

 		2.2.1 Date::Parse
		2.2.2 Net::DNS
		2.2.3 Digest
		2.2.4 Digest::HMAC
		2.2.5 Digest::MD5
		2.2.6 DBI
		2.2.7 Encode::MIME::Header

3. Features

	3.1 A higly configurable banlist (see banlist.conf(5))

	3.2 DNSBL, URIBL and TOR support

	3.3 The maximum number of

		3.3.1 Messages
		3.3.2 Bytes
		3.3.3 Groups
		3.3.4 Followups
		3.3.5 Identical messages
		3.3.6 Rejected messages before being banned

    		can be limited per

		3.3.7 UserID
		3.3.8 Domain
		3.3.9 IP address

    		and per *two* units of time (by default: 1 Day and 10 Minutes)

	3.4 Headers can be deeply modified

	3.5 Builtin log system for data that need to be kept for legal reasons (notably sender's IP address of each locally posted message) with
	    an expernal script (postlegal that is included in this package) which handles these logs (expiration of old logs and searches)

	3.6 A simple and fast dictionary of forbidden words (see badwords.conf)

	3.7 Easy to customize

4. Installation

	4.1 Automatic installation:
	    
	    Run as root

		$ rc.news stop		  # Stop INND
		$ ./postfilter-install -c # Check dependacies
		$ ./postfilter-install -i # Install postfilter
		$ rc.news start		  # Start INND

	4.2 Installation by hands

	    Read ./doc/tutorials/how_to_install_postfilter_by_hands.txt

5. Troubleshooting

	5.1 Enable news.debug syslog facility adding the following line in syslog.conf

#-----------------------------------------------------------------
news.debug                      -/var/log/news/news.debug
#-----------------------------------------------------------------

	5.2 Check whether postfilter is correctly installed:

        	$ perl -wc [pathfilter]/postfilter/postfilter.pl 
        	$ ./postfilter-installer -c
        	$ su root
        	$ ./postfilter-installer -v
	
	5.3 If installer complaints because it doesn't find innconfval, 
	    which is needed during installation, its full path has to be 
	    inserted by hands using -s flag
	    
	    	$ su root
		$ updatedb
		$ locate innconfval
		  /news/bin/innconfval
		$ postfilter-installer -i -s /news/bin/innconfval

	5.4 Restart INND. Please *restart* it instead of reloading the configuration settings.

        	$ su root
        	$ rc.news stop
        	$ rc.news start

	5.5 Try to post a test message

		$ telnet localhost 119
              >>> Trying 127.0.0.1...
	      >>> Connected to localhost.
	      >>> Escape character is '^]'.
	      >>> 200 aioe.org InterNetNews server INN 2.4.5 ready
	      <<< MODE READER
	      >>> 200 localhost InterNetNews NNRP server INN 2.4.5 ready (posting ok).
	      <<< POST
	      >>> 340 Ok, recommended ID <ha550k$be9$1@aioe.org>
	      <<< Newsgroups: local.test                           
	      <<< From: Aioe <estasi@aioe.org>
	      <<< Subject: test
	      <<<
	      <<< This is a test, only a test
	      <<<.
	      >>> 240 Article posted <ha550k$be9$1@aioe-casa.aioe.org>


	5.6 Read all news.debug informations related with postfilter

        	$ tail -f /var/log/news/news.debug | grep " filter: "

	5.7 If postfilter complaints about configuration errors, read

		$ man postfilter
		$ man postfilter.conf
		$ man banlist.conf	


	5.8 Those who need help can ask for assistance using:

        	* news://news.software.nntp (an USENET newsgroup)
        	* news://nntp.aioe.org/aioe.news.helpdesk (a local group)



CHANGELOG
Postfilter 0.9.1

Enhancements:
	1. Postfilter requires also Encode::MIME::Header (postfilter.pl and postfilter-installer were updated)

Bugfixes:

	1. If the argument of From or Subject header includes a character that is forbidden, argument is converted to MIME format
	2. In order to check if a user has the right to post the message that is currently checked, postfilter now sums the values 
	   ​estracted from the current message to those related to messages already sent in the past. 
	3. Removed useless subroutines inside postfilter.pl
	4. If postfilter is configured in order to save accepted and/or refused local messages, each post is saved in a file inside
           [pathspool]/postfilter/saved/ using a name with format [3 digit error code].[UNIX timestamp] if it was rejected or 
	   00.[UNIX Timestamp] if it was accepted.

Postfilter 0.8.3

Enhancements:

	1. full support to INN 2.6
	2. A syntax checker for configuration files
	3. Installer allows to set innconfval path by hands
	4. Added $config{"allow_html"} that sets whether a message with HTML tags is rejected
	5. Distribution header is supported (Distribution: local|usenet)

Bugfixes:

	1. URIBLs work as expected
	2. Cache bugfix
	3. Installer works as expected
	4. Content-Type check now works correctly
	5. Invalid regexps in banlist.conf and badwords.conf now don't crash the filter
	6. Messages with non standard chars in the argument of an header are rejected
	


Postfilter 0.8.1

Enhancements:
	
	1. A better installer
	2. doc/ is now more complete
	3. postlegal is able to generate statistical data


Bugfixes:
 
       1. Postfilter works with INN 2.5	and 2.6








src/modules/other.pm
# Copyright (c) 2005-2019, Paolo Amoroso (Aioe) <freedom@aioe.org>
# All rights reserved.
# Postfilter (other.pm) version 0.9.1

use Date::Parse;
use Digest::MD5;
use Digest::SHA1;
use Digest;
use strict;

our (%hdr, $dbh, @access, $modify_headers, $body, $user, %config, %public_rights_ip, %public_rights_domain, %auth_rights, %ban_limits);
our (%headlist, @quickref, @saved_headers, %whitelist, @distributions, %mysql, %forbidden_crosspost, %scoreset, %maxscore);
our (@nomoderation, @htmlallowed, @htmltags, %extracontent, %forbidden_headers, @forbidden_groups,%dnsbl,@localip,%attributes);

########
#
# create_cache_header(): Create header Cache
#
#######


sub create_cache_header()
{

	my ( $host, $time, $pid, $ip, $date, $client_dn );

	if (($config{'version'} eq "2.4") or ($config{'version'} eq "2.5"))
	{
        	( $host, $time, $pid, $ip, $date ) = split( /\ /, $hdr{'X-Trace'}, 5 );
		$ip = "127.0.0.1" if ( $ip =~ /\:\:1/ ); # This will be removed as soon as IPv6 support will become stable
        	$client_dn = $hdr{'NNTP-Posting-Host'};
	} else {

		$time = time();
		$client_dn = $attributes{'hostname'};
		$ip = $attributes{'ipaddress'};
		$host = $attributes{'interface'};
		$date = $hdr{'Injection-Date'};
		$pid = 999; #fixme
	}


########################
# Body Length
########################

        my $length = length($body);

########################
# Headers Length
########################

        my $temp_headers = "";

        foreach ( keys %hdr )
        {
		$temp_headers .= "$_\: $hdr{$_}\n";
        }

        my $head_length = length( $temp_headers );

########################
# Article MD5
########################

        my $md5article;

        if ( $config{'md5_hash'} eq "body" )
        {
                $md5article = $body;
        }
        elsif ( $config{'md5_hash'} eq "simple" )
        {
                $md5article = "Subject: $hdr{'Subject'}\n\n" . $body;
        }
        elsif ( $config{'md5_hash'} eq "complex" )
        {
                $md5article = "Newsgroups: $hdr{'Newsgroups'}\n" . "Subject: $hdr{'Subject'}\n\n" . $body;
        }
        elsif ( $config{'md5_hash'} eq "all" )
        {
                $md5article = "Newsgroups: $hdr{'Newsgroups'}\n" . "From: $hdr{'From'}\n" . "Followup-To: $hdr{'Followup-To'}\n" . "Subject: $hdr{'Subject'}\n\n" . $body;
        }
        else
        {
                &log( "err", "Invalid criteria ($config{'md5_hash'}) for MD5 hash, using \"body\" ");
                $md5article = $body;
        }

        my $md5 = &create_md5($md5article);

########################
# Client's domain
########################

        my $domain = "";
        my @temp_domain = split( /\./, $client_dn );

        my $dots = 0;

        while ( $client_dn =~ /\./g )                    # debug
        {
                $dots++;
        }

	
        if ( $dots > 1 )                                        # se ci sono almeno due punti (e quindi TRE domini) ne
        {                                                       # togle uno, altrimenti copia l'host in domain
                shift( @temp_domain ) if ( $ip ne $client_dn ); # se e' un dominio cancella la parte a sinistra
                pop  ( @temp_domain ) if ( $ip eq $client_dn ); # se e' un ip elimina la parte a destra

                for ( my $n = 0; $temp_domain[$n] ne ""; $n++ )
                {
                        $domain .= $temp_domain[$n] . ".";
                }

                chop $domain;
        }
        else
        {
                $domain = $client_dn;
        }

########################
# Crosspost & Followup
########################

        my @gruppi   = split( /[, ]+/, $hdr{'Newsgroups'}  );
        my @followup = split( /[, ]+/, $hdr{'Followup-To'} );

        my $gruppi = @gruppi;
        my $followup = @followup;

        $modify_headers = 1;
        $hdr{'Cache'} = "$host $time $pid $ip $client_dn $length $head_length $md5 $domain $gruppi $followup";


        return 0;
}

########
#
# query_cache_header(): Query header Cache
#
#######

sub query_cache_header()
{
        my $text = $hdr{'Cache'};
        my @data = split( /\ /, $text );
        return @data;
}

########
#
# delete_cache_header(): Delete header Cache
#
#######

sub delete_cache_header()
{
	$modify_headers = 1;
	&delete_headers('Cache');
	return 0;
}

########
#
# save_message(): Save a message in a file
#
#######

sub save_message()
{
        my ($host, $time, $pid, $ip, $client_dn, $length, $head_length, $md5, $domain, $gruppi, $followup) = &query_cache_header();

        my $file_name;

########################
# Filename
########################

        if ( $hdr{'X-Postfilter-Error'} ne "" )
        {
                $file_name = sprintf( "%03d\.%d", $hdr{'X-Postfilter-Error'}, $time );
        }
        else
        {
                $file_name = "00" . "." . $time;
        }

########################
# File Path
########################

        my $file_path = $config{'dir_spool'} . "/saved/" . $file_name;

        &log( "notice", "Saving $hdr{'Message-ID'} into $file_path" );

########################
# Save
########################

        open my $SAVE, ">$file_path";
        if (!$SAVE)
        {
                &log( "err", "Unable to save message into $file_path" );
                return( 41 );
        }

        foreach (keys %hdr)
        {
                print $SAVE "$_: $hdr{$_}\n";
        }

        print $SAVE "\n$body\n.\n";
        close $SAVE;
        return 0;
}

########
#
# custom_filter(): Run custom rules (custom.pm)
#
#######

sub custom_filter()
{
	my $error_code;

########################
# Run customized rules
########################

	eval { $error_code = &custom_rules() };
	
	if ($@)
	{
		&log( "err", "Unable to execute custom rules: $@" );
		return 0;
	}

########################
# Errors handling
########################

	if ($error_code =~ /^[0-9]+/)
	{
		if ($quickref[$error_code] ne "" )
		{
			return $error_code;
		}		
		else
		{
			&log("err", "Custom rule returns $error_code and the corresponding error string is black (\'$quickref[$error_code]\': user will receive a generic return string");
			$quickref[$error_code] = $quickref[63];
			return $error_code;
		}
	}
	else
	{
		&log( "err", "Custom rules return a bad value: \'$error_code\' instead of a number, message accepted" ); 
		return 0;
	}

	return 0;
}

########
#
# create_md5(): Create MD5 sign of a string
#
#######

sub create_md5($)
{
	my $text = $_[0];
        my $ctx = Digest::MD5->new;
        $ctx->add($text);
        my $md5 = $ctx->b64digest;
	return $md5;
}

########
#
# check_uuencode(): Check whether an article include UUEncoded text
# 
#######


sub check_uuencode()
{
	my @body = split(/\n/, $body );
        my $stat = 0;
        foreach ( @body )
        {
 	       $stat = 1 if ($_ =~ /^begin\ [0-9][0-9][0-9]/i  );
               $stat = 2 if (($_ =~ /^end/i) and ( $stat == 1 ));
        }

	if ( $stat == 2 )
        {
                return 52;
        }
	return (0);
}

########
#
# check_yenc(): Check whether an article include Yenc text
#
#######


sub check_yenc()
{
        my @body = split(/\n/, $body );
        my $stat = 0;
        foreach ( @body )
        {
               $stat = 1 if ($_ =~ /^\=ybegin\ /i);
               $stat = 2 if (($_ =~ /^\=yend/i) and ( $stat == 1 ));
        }

        if ( $stat == 2 )
        {
                return 62;
        }
        return (0);
}

########
#
# delete_headers(@headers_to_delete): Delete one o more headers
#
#######


sub delete_headers($)
{
        $modify_headers = 1;
        my @headers = @_;
        my $system_headers =    "^Path|^From|^Newsgroups|^Subject|^Control|^Supersedes|" .
                                "^Followup\-To|^Date|^Organization|^Lines|^Sender|^Approved|^Distribution|" .
                                "^Expires|^Message\-ID|^References|^Reply\-To|^NNTP\-Posting\-Host|" .
                                "^Mime\-Version|^Content\-Type|^Content\-Transfer\-Encoding|^X\-Trace|" .
                                "^X\-Complaints\-To|^NNTP\-Posting\-Date|^Xref|^Injector\-Info|^Summary|" .
                                "^Keywords|^Date\-Received|^Received|^Posted|^Posting\-Version|" .
                                "^Relay\-Version|^Bcc|^To|^CC";

        foreach ( @headers )
        {
                if ( $hdr{$_} ne "" )  # check whether that header exists
                {

########################
# Delete headers with 2.4
########################

                        if ( $config{'version'} eq "2.4" )
                        {
                                if ( $_ =~ /$system_headers/i )
                                {
                                        $hdr{$_} = undef;
                                        &log( "debug", "System header $_ set to undef" );
                                }
                                else
                                {
                                        delete $hdr{$_};
                                        &log( "debug", "Optional header $_ deleted" );
                                }
                        }

########################
# Delete headers with 2.5 or 2.6
########################
                        else
                        { 
                                $hdr{$_} = undef;
                        }
                }
                else
                {
                        &log( "debug", "Detected an attempt to delete $_ that doesn't exist" );
                }

        }
}


########
#
# read_active(): Read active file
#
#######

sub read_active()
{
	open my $ACTIVE, "$config{'file_active'}";
        if (!$ACTIVE)
        {
                &log( "err", "Unable to read data from $config{'file_active'}" );
                return 37;
        }
        my @active = <$ACTIVE>;
        close $ACTIVE;

	return @active;
}

########
#
# check_group_existence($group): Check whether a group exists
#
#######

sub check_group_existence($)
{
	my $group = $_[0];

	my @active = &read_active();

	foreach ( @active )
        {
        	my ($group_string, $first, $last, $type) = split( /\ /, $_ );
                return 1 if ( $group_string eq $group );
        }

	return 0;

}

sub checkuserdb()
{
	my $grouptocheck = shift(@_);

	&log("notice", "Message sent to a group ($grouptocheck) protected by userdb");

        my $FILEID;
        my $fileid = "/etc/news/postfilter/userdb/$grouptocheck.dat";

        $quickref[105] = "Unable to load $fileid";
        
	open $FILEID, "$fileid" or return 105;
        my @fromlist = <$FILEID>;
        close $FILEID;

        foreach (@fromlist)
        {
	        my $ffff = $hdr{'From'};
                my $gggg = $_;
                chop($gggg);
                if ($gggg eq $ffff)
                {
        	        &log("notice", "Article sent by a known user: $gggg");
                        return 0;       
                }
	}
        
	$quickref[106] = "Non puoi postare su $grouptocheck, contatta usenet\@aioe.org";
        my $ffgg = $hdr{'From'};

        &log("err", "Message sent by an unknown user: $ffgg");

        return 106;
}


sub save_localdistribution()
{
        my $mid = shift(@_);
        my $news_spool = $config{'dir_spool'} . "/localdistribution.dat";
        my $expire_time = 86400 * 90;
        my $current_time = time();

	$quickref[107] = "Unable to write $news_spool";

        if (open(FILE, "$news_spool"))
        {
                my @lines = <FILE>;
                close FILE;

                if (open(FILE, ">", $news_spool))
                {
                
                        foreach(@lines)
                        {
                                my ($epoch,$oldmid) = split(/\t/, $_);
                                print FILE "$epoch\t$oldmid" if ( $epoch > $current_time - $expire_time );
                        }
                                
                        print FILE "$current_time\t$mid\n";
                        close FILE;             
                } else {
                        &log("err", "Impossibile scrivere su $news_spool");       
                        return 107;
                }
        } else {
                &log("err", "Impossibile aprire $news_spool");
                return 107;
        }

        return 0;
}


sub check_localdistribution()
{
        my $mid = shift(@_);
        my $news_spool = $config{'dir_spool'} . "/localdistribution.dat";

        $quickref[107] = "Unable to write $news_spool";

        if (open(FILE, "$news_spool"))
        {
                my @lines = <FILE>;
                close FILE;

                foreach(@lines)
                {
                                my ($epoch,$oldmid) = split(/\t/, $_);
				chop $oldmid;
				return 1 if ($mid eq $oldmid);
                }
        } else {
                &log("err", "Unable to open $news_spool");
                return 107;
        }

        return 0;
}

sub check_unknownchars()
{
	my $head = shift(@_);
 	my $string = $hdr{$head};
        my $DDposition = 1;   # conta l'ennesimo carattere quindi parte da uno

        my @ASCII = unpack("C*", $string);

	&log("debug", "Checking header $head againt invalid chars");

        foreach (@ASCII)
        {
                if (
                        (( $_ >= 32 ) && ( $_ <=  90 )) or
                        (( $_ >= 94 ) && ( $_ <= 126 )) or
                        (  $_ == 195 ) or
			(  $_ == 185 ) or
			(  $_ == 178 ) or
			(( $_ >=   9 ) && ( $_ <=  10 )) or
			(  $_ ==  91 ) or
			(  $_ ==  93 )
                   )
                {
                        $DDposition++;
                }
                else
                {
                        $quickref[104] = "Forbidden ($DDposition th) char (ASCII $_) inside $head";
			&log("err", "Header: $head\: $string .. Invalid $DDposition th char ASCII $_");
                        return 104;
                }
        }

	return 0;
}

	
1;
